### PR TITLE
[pruning][core][feature] Align BaseStructuredPruner with existing pruning flow

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89777
* __->__ #88436

Summary:

This PR aligns the "eager" mode of the structured pruning flow with the existing unstructured pruning flow.

The base pruner has been moved from and has been renamed from BasePruner to BaseStructuredPruner
`torch/ao/pruning/_experimental/pruner/base_pruner.py -> /torch/ao/pruning/_experimental/pruner/base_structured_pruner.py`

Support for pruning batchnorm modules in the config have been removed, so now the structured pruning code can use more of the BaseSparsifier logic and we don't need to override as many functions.

Since we aim to only support a single flow, we have only updated ZeroesParametrizations (FakeStructuredSparsity) and BiasHook.
The parameterizations have also been rewritten to use a bool mask tensor for keeping track of pruned rows, instead of using sets before.
This better aligns structured and unstructured sparsity.

The BaseStructuredSparsifier tests have also been updated to reflect the above changes. I also removed `squash_mask` tests because they were breaking CI and `squash_mask` is no longer used.

We will migrate the structured pruning code out of this folder in a later PR.

Test Plan:
```
python test/test_ao_sparsity -- TestBaseStructuredPruner
```

Reviewers:
z-a-f vkuzo

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: test/test_ao_sparsity.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 from ao.sparsity.test_sparsifier import TestNearlyDiagonalSparsifier  # noqa: F401
 
 # Pruner
-from ao.sparsity.test_pruner import TestBasePruner  # noqa: F401
+from ao.sparsity.test_pruner import TestBaseStructuredPruner  # noqa: F401
```
Comment (vkuzo) [inline test/test_ao_sparsity.py:19]: should we move `test/ao/sparsity` to `test/ao/pruning`, and rename the tests?  cc @z-a-f , ideally that would be a separate PR

File: torch/ao/pruning/sparsifier/base_structured_pruner.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+from .utils import FakeStructuredSparsity, BiasHook
+
+__all__ = ["BaseStructuredPruner"]
+
+SUPPORTED_STRUCTURED_PRUNING_MODULES = {  # added to config if None given
+    nn.Linear,
+    nn.Conv2d,
+}
+
+class BaseStructuredPruner(BaseSparsifier):
```
Comment (vkuzo) [inline torch/ao/pruning/sparsifier/base_structured_pruner.py:16]: we should keep the naming consistent, if the parent is `BaseSparsifier` the child should be `BaseStructuredSparsifier`.  The name change should be a separate PR.

File: torch/ao/pruning/sparsifier/base_structured_sparsifier.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+            prune_bias = config.get("prune_bias", True)
+            if prune_bias and module.bias is not None:
+                module.register_parameter("_bias", nn.Parameter(module.bias.detach()))
+                module.bias = None
+            self.bias_handles.append(
+                module.register_forward_hook(
+                    BiasHook(module.parametrizations.weight[0], prune_bias)
+                )
+            )
```
Comment (vkuzo) [inline torch/ao/pruning/sparsifier/base_structured_sparsifier.py:70]: can we check if this survives symbolic tracing?  For example, if a user applies `BaseStructuredSparsifier`, then symbolically traces the model with FX, and then runs forward, will the bias hook be applied?

**Review by vkuzo — APPROVED** [review]
thanks for working on this!

File: test/test_ao_sparsity.py (hunk 1 above)
Comment (z-a-f) [reply test/test_ao_sparsity.py]: For consistency, I think it would make sense to rename the test folder.

File: torch/ao/pruning/sparsifier/base_structured_pruner.py (hunk 2 above)
Comment (z-a-f) [reply torch/ao/pruning/sparsifier/base_structured_pruner.py]: I think we should take a step further, and rename all to "*Pruner". The reason we had `Sparsifier` and `Pruner` was because of the outdated definitions: "sparity" was ="unstructured", while "pruning was ="structured". But now those definitions are not applicable anymore. Could be in future PRs though

**Comment by z-a-f** [comment]
Would it be possible to split this into two separate diffs: (1) Changing the code / renaming the classes, and (2) Moving out of the `_experimental`. That way we would preserve the blame history and avoid accidental bugs creeping in.

**Comment by z-a-f** [comment]
Is there a reason for changing the sets into masks? That will incur significant memory penalty. Also, what is the shape of the mask when a many-dimensional weight is being pruned?

**Comment by jcaip** [comment]
> Is there a reason for changing the sets into masks? That will incur significant memory penalty. Also, what is the shape of the mask when a many-dimensional weight is being pruned?

I thought it would better to keep the structured and unstructured pruning aligned. It also makes some of the bias propagation / fusion logic cleaner. 

Can you explain the significant memory penalty a bit more? We're not creating a mask of the size of the weight tensor, just the size of the # of rows. Then we use tensor broadcasting when multiplying the mask by the weight. 

So for a linear weight of size (R, C), our mask is of size (R) and for a conv layer weight of size (C_out, C_in, H, W) our mask is of size (C_out). 

[Before](https://github.com/pytorch/pytorch/blob/cdb798faefa2520b37938311bcef1c175581a0ff/torch/ao/pruning/_experimental/pruner/parametrization.py#L7) we were keeping two sets, one of the original outputs, and one of the pruned outputs. So in the linear case of size (R, C), this is a set of integers of size R (for the original outputs) and then an additional set of integer of some size (for the pruned elements). So we have strictly > R elements in the sets, which would take up more space than our mask of just size R. 

Not to mention the elements in the set are ints vs bools for the mask.

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+__all__ = ["BaseStructuredSparsifier"]
+
+SUPPORTED_STRUCTURED_PRUNING_MODULES = {  # added to config if None given
+    nn.Linear,
+    nn.Conv2d,
+}
+
+
+class BaseStructuredSparsifier(BaseSparsifier):
```
Comment (z-a-f) [inline torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:17]: Technically there is no such thing as "Structured Sparsity". Although most readers would understand that this means "structured pruning", if we want to be consistent, we should not use this name
